### DISCUSSION TITLE
Need for a switch-to-MIDI converter

### DISCUSSION THREAD

**Discussion by jason-barba** [GrandOrgue/grandorgue#1836] [Help]
Currently I run MIDI from manuals and pedal via MIDI thru a FC 1010 foot-controller to a MIDI-USB converter.

I now see the need for some additional general pistons (minimum 2). The switch-to-midi converter has to have MIDI-thru; I have no more free USB ports, so another MIDI-USB converter is out of question.

What device can fullfill my need? Any suggestion is highly welcome.

//Jens

**Comment by midimusic** [comment]
You can combine multiple usb ports into one port which feeds your computer. Check out "usb combiner" on line and you will see multiple choices, none of which are expensive

csw900

**Reply by Aduragbemi-Adeyinka** [reply]
Maybe a **powered** USB hub will help.

**Comment by davidgritter** [comment]
you can use a raspberry pi pico for this at about $7.00. It can be programmed to create a MIDI in on its serial1 port -you need to add an external optocoupler to do a true midi input, there's lots of examples on the internet.  With the earle philhower pico board package, and the adafruit tiny usb library you can turn the pico usb connector into a a usb midi port to connect to your grandorgue computer.  The midi inputs that come to your midi to usb device now would go to the pico midi input.  The attached arduino sketch shows the code to implement the midi passthrough, sample to expression pedals, and interface to lighted pushbuttons. You probably can eliminate the lights for the pushbuttons  and the expression pedal code pretty easily.
Alternatively you could use the USB hub mentioned above and avoid the need for midi passthrough. 

 [Rodgers_to_USB_pico.zip](https://github.com/GrandOrgue/grandorgue/files/14566402/Rodgers_to_USB_pico.zip)

**Reply by jason-barba** [reply]
Thanks a plenty.

If I understand correctly, your code outputs MIDI-messages on a USB port. Correct?

I am a bit confused. Your code is a .ino file implementing Arduino but your hardware suggestion is a Raspberry Pi. Please enlighten me.

**Reply by davidgritter** [reply]
Its a raspberry pi PICO, small inexpensive board ($7) (see picture) that can be programmed with arduino.  
![pico](https://github.com/GrandOrgue/grandorgue/assets/94871928/b4e8a5b3-191b-40eb-98c2-e19abcaff5c0)

Earle Philhower makes a PICO board support package for the arduino IDE that works very well for midi applications, and the code I sent was for that board support package. see https://arduino-pico.readthedocs.io/en/latest/.  I use the Arduino IDE 2 which allows hardware debugging of the PICO using a second PICO board with a special debug program provided by the Raspberry pi people.

The PICO has native USB and the philhower package supports the ADAFRUIT TINY_USB packages which can be configured to include USB MIDI.    The Serial port on  PICO pins 1 and 2 can be configured as a standard midi port for input and/or output to standard MIDI cables.
The USB port can simultaneously support midi and a serial port to make debugging easier.

**Reply by jason-barba** [reply]
Thank you! :-)
